Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm

```diff
@@ -439,11 +439,44 @@ static inline String toMIMEType(WebExtensionTab::ImageFormat format)
     ASSERT(contentScriptProcesses.size() == 1);
     auto process = contentScriptProcesses.takeAny();
 
-    process->sendWithAsyncReply(Messages::WebExtensionContextProxy::DispatchRuntimeContentScriptMessageEvent(messageJSON, frameIdentifier, senderParameters), [protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler)](std::optional<String> replyJSON) mutable {
+    process->sendWithAsyncReply(Messages::WebExtensionContextProxy::DispatchRuntimeMessageEvent(WebExtensionContentWorldType::ContentScript, messageJSON, frameIdentifier, senderParameters), [protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler)](std::optional<String> replyJSON) mutable {
         completionHandler(replyJSON, std::nullopt);
     }, identifier());
 }
 
+void WebExtensionContext::tabsConnect(WebExtensionTabIdentifier tabIdentifier, WebExtensionPortChannelIdentifier channelIdentifier, String name, std::optional<WebExtensionFrameIdentifier> frameIdentifier, const WebExtensionMessageSenderParameters& senderParameters, CompletionHandler<void(WebExtensionTab::Error)>&& completionHandler)
+{
+    // Add 1 for the starting port here so disconnect will balance with a decrement.
+    constexpr auto sourceContentWorldType = WebExtensionContentWorldType::Main;
+    addPorts(sourceContentWorldType, channelIdentifier, 1);
+
+    auto tab = getTab(tabIdentifier);
+    if (!tab) {
+        completionHandler(toErrorString(@"tabs.connect()", nil, @"tab not found"));
+        return;
+    }
+
+    constexpr auto targetContentWorldType = WebExtensionContentWorldType::ContentScript;
```

Comment:
If the tab we are trying to connect to is showing full-page extension content, would this still be content script? Or in that case would it be main world?

Or in that case, are extension developers not supposed to use `tabs.connect`?

Reply:
Correct, `runtime.connect` is used for full page content messaging in the main world (and with popups). `tabs.connect` is only for content scripts.